telethon_examples/payment.py

```diff
@@ -0,0 +1,174 @@
+    while True:
+        value = input(message)
+        try:
+            return cast(value)
+        except ValueError as e:
+            print(e, file=sys.stderr)
+            time.sleep(1)
+
+
+bot = TelegramClient(
+    os.environ.get('TG_SESSION', 'payment'),
+    get_env('TG_API_ID', 'Enter your API ID: ', int),
+    get_env('TG_API_HASH', 'Enter your API hash: '),
+    proxy=None
+)
+
+
+# That event is handled when customer enters his card/etc, on final pre-checkout
+# If we don't `SetBotPrecheckoutResultsRequest`, money won't be charged from buyer, and nothing will happen next.
+@bot.on(events.Raw(types.UpdateBotPrecheckoutQuery))
+async def payment_pre_checkout_handler(event: types.UpdateBotPrecheckoutQuery):
+    if event.payload.decode("UTF-8") == 'product A':
+        #  so we have to confirm payment
+        await bot(
+            functions.messages.SetBotPrecheckoutResultsRequest(
+                query_id=event.query_id,
+                success=True,
+                error=None
+            )
+        )
+    elif event.payload.decode("UTF-8") == 'product B':
+        # same for another
+        await bot(
+            functions.messages.SetBotPrecheckoutResultsRequest(
+                query_id=event.query_id,
+                success=True,
+                error=None
+            )
+        )
+    else:
+        # for example, something went wrong (whatever reason). We can tell customer about that:
+        await bot(
+            functions.messages.SetBotPrecheckoutResultsRequest(
+                query_id=event.query_id,
+                success=False,
+                error="Something went wrong"
+            )
+        )
+
+    raise events.StopPropagation
+
+
+# That event is handled at the end, when customer payed.
+@bot.on(events.Raw(types.UpdateNewMessage))
+async def payment_received_handler(event):
+    if isinstance(event.message.action, types.MessageActionPaymentSentMe):
+        payment: types.MessageActionPaymentSentMe = event.message.action
+        # do something after payment was recieved
+        if payment.payload.decode("UTF-8") == 'product A':
+            await bot.send_message(event.message.from_id, "Thank you for buying product A!")
+        elif payment.payload.decode("UTF-8") == "product B":
+            await bot.send_message(event.message.from_id, "Thank you for buying product B!")
+        raise events.StopPropagation
+
+
+# let's put it in one function for more easier way
+def generate_invoice(price_label: str, price_amount: int, currency: str, title: str,
+                     description: str, payload: str, start_param: str) -> types.InputMediaInvoice:
+    price = types.LabeledPrice(label=price_label, amount=price_amount)  # label - just a text, amount=10000 means 100.00
+    invoice = types.Invoice(
+        currency=currency,  # currency like USD
+        prices=[price],  # there could be a couple of prices.
+        test=True,  # if you're working with test token

```

Comment:
I don't see any explanation on how someone would work with non-test data. How do I even get started with real payments? I think we should at least include an explicit link to https://core.telegram.org/bots/payments, in both the `README` and top of the file:
```python
"""
blah blah https://core.telegram.org/bots/payments blah blah to get started…
…make sure to change `test=False` below when done blah blah
"""
```

So that people won't ask "okay but how do I actually use real data?".

Reply:
Provider token is needed both for test and real though.

Reply:
Hmm, okay, I still think a brief introduction to the matter would be nice. The line:
```python
 provider_token = ""  # https://core.telegram.org/bots/payments#getting-a-token 
```

May be missed and we don't check if the user didn't provide one, which likely fails with some RPC error.

Reply:
Added comments with a new commit